src/chain/sync/grandpa_warp_sync.rs

```diff
@@ -0,0 +1,288 @@
+    WarpSyncRequest(WarpSyncRequest<'a>),
+    /// Fetching the parameters for the virtual machine is required to continue.
+    VirtualMachineParamsGet(VirtualMachineParamsGet<'a>),
+}
+
+impl<'a> GrandpaWarpSync<'a> {
+    fn from_babe_fetch_epoch_query(
+        query: babe_fetch_epoch::Query,
+        fetched_current_epoch: Option<BabeEpochInformation>,
+        state: PostVerificationState<'a>,
+    ) -> Self {
+        match (query, fetched_current_epoch) {
+            (babe_fetch_epoch::Query::Finished(Ok((next_epoch, runtime))), Some(current_epoch)) => {
+                let slots_per_epoch = match state.genesis_chain_information.consensus {
+                    ChainInformationConsensus::Babe {
+                        slots_per_epoch, ..
+                    } => slots_per_epoch,
+                    _ => unreachable!(),
+                };
+
+                Self::Finished(Ok((
+                    ChainInformation {
+                        finalized_block_header: state.header,
+                        finality: state.chain_information_finality,
+                        consensus: ChainInformationConsensus::Babe {
+                            finalized_block_epoch_information: Some(current_epoch),
+                            finalized_next_epoch_transition: next_epoch,
+                            slots_per_epoch,
+                        },
+                    },
+                    runtime,
+                )))
+            }
+            (babe_fetch_epoch::Query::Finished(Ok((current_epoch, runtime))), None) => {
+                let babe_next_epoch_query =
+                    babe_fetch_epoch::babe_fetch_epoch(babe_fetch_epoch::Config {
+                        runtime,
+                        epoch_to_fetch: babe_fetch_epoch::BabeEpochToFetch::NextEpoch,
+                    });
+                Self::from_babe_fetch_epoch_query(babe_next_epoch_query, Some(current_epoch), state)
+            }
+            (babe_fetch_epoch::Query::Finished(Err(error)), _) => {
+                Self::Finished(Err(Error::BabeFetchEpoch(error)))
+            }
+            (babe_fetch_epoch::Query::StorageGet(storage_get), fetched_current_epoch) => {
+                Self::StorageGet(StorageGet {
+                    inner: storage_get,
+                    fetched_current_epoch,
+                    state,
+                })
+            }
+            (babe_fetch_epoch::Query::NextKey(next_key), fetched_current_epoch) => {
+                Self::NextKey(NextKey {
+                    inner: next_key,
+                    fetched_current_epoch,
+                    state,
+                })
+            }
+        }
+    }
+}
+
+/// Loading a storage value is required in order to continue.
+#[must_use]
+pub struct StorageGet<'a> {
+    inner: babe_fetch_epoch::StorageGet,
+    fetched_current_epoch: Option<BabeEpochInformation>,
+    state: PostVerificationState<'a>,
+}
+
+impl<'a> StorageGet<'a> {
```

Comment:
You need to indicate which `PeerId` we have downloaded the Grandpa warp sync from. Same in `NextKey`.
In practice, these `StorageGet` requests will be fulfilled by asking the peer as well.